extras/create_send_notifications.py

```diff
@@ -42,26 +42,41 @@ def create_output_directories():
         os.path.join(NOTIFICATIONS_BASE_DIR, NOTIFICATION_ARCHIVE_DIR)
     )
 
+def build_notifications_org_list(db):
+    """Return notifications organization list.
 
-def build_cyhy_org_list(db):
-    """Build list of CyHy organization IDs.
-
-    This is the list of CyHy organization IDs (and their descendants) that
-    receive CyHy reports.
+    This is the list of organization IDs that should
+    have a notification generated and sent.
+    """
+    notifications_to_generate = set()
+    cyhy_parent_ids = set()
+    ticket_owner_ids = db.notifications.distinct("ticket_owner")
+    for request in db.RequestDoc.collection.find({"_id": {"$in": ticket_owner_ids}, "report_types": "CYHY"}, {"_id":1}):
+        # If the notification document's ticket owner has "CYHY" in their list of report_types,
+        # then a notification should be generated for that owner:
+        notifications_to_generate.add(request["_id"])
+        # Recursively check for any parents of the ticket owner that have "CYHY" in
+        # their list of report_types.  If found, add them to the list of owners that
+        # should get a notification.
+        cyhy_parent_ids = cyhy_parent_ids | find_cyhy_parents(db, request["_id"])
+    notifications_to_generate.update(cyhy_parent_ids)
+    notifications_to_delete = set(ticket_owner_ids) - notifications_to_generate
+    return list(notifications_to_generate), list(notifications_to_delete)
+          
+def find_cyhy_parents(db, org_id):
+    """Return parents/grandparents/etc. of an organization that have "CYHY" in their list of report_types.
     """
-    cyhy_org_ids = set()  # Use a set here to avoid duplicates
-    for cyhy_request in list(
-        db.RequestDoc.collection.find(
-            {"report_types": "CYHY"}, {"_id": 1, "children": 1}
-        ).sort([("_id", 1)])
-    ):
-        cyhy_org_ids.add(cyhy_request["_id"])
-        if cyhy_request.get("children"):
-            cyhy_org_ids.update(db.RequestDoc.get_all_descendants(cyhy_request["_id"]))
-    return list(cyhy_org_ids)
-
-
-def generate_notification_pdfs(db, org_ids, master_report_key):
+    cyhy_parents = set()
+    for request in db.RequestDoc.collection.find({"children": org_id, "report_types": "CYHY"}, {"_id": 1}):
+        # Found a parent of org_id with "CYHY" in their list of report_types,
+        # so add it to our set
+        cyhy_parents.add(request["_id"])
+        # Recursively call find_cyhy_parents() to check if this org has any parents
+        # with "CYHY" in their list of report_types
+        cyhy_parents.update(find_cyhy_parents(db, request["_id"]))
```

Comment:
@dav3r you're generally more familiar with customer structures. Would it be possible to have an org that receives `CYHY` reports, a parent org that does not, and a grandparent (specifically a parent of the org **not** receiving) that also receives `CYHY` reports? I ask because this logic looks to stop searching upward (for a given parent) as soon as a parent in the chain is **not** receiving `CYHY` reports.

Reply:
Given that the constraint is undocumented, following the hierarchy to the top is probably the best solution as long as it doesn't bog things down too much.  I'd also add a comment stating that there is an undocumented constraint, so we don't expect to ever need to go up more than one level, but we are doing this to be safe.  That way we'll remember all this should we ever revisit this code.

Reply:
@mcdonnnj ; @jsf9k  Please see https://github.com/cisagov/cyhy-reports/pull/96/commits/f616f3b29ff98c48ed80181f56654a5293f9a61b